### PR TITLE
[JENKINS-67173] Deprecate `SlaveToMasterFileCallable`, log warning

### PR DESCRIPTION
**Pull request by daniel-beck**
Start telling people that S2MFCs are bad. There's a reason we need S2MCs (no `File`) for very specific, narrow circumstances, but S2MFCs (and really, most S2MCs) need to go.

Untested.

### Proposed changelog entries

* Deprecate `SlaveToMasterFileCallable`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/SlaveToMasterFileCallable.java
Comment (jglick) [inline core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]:
```suggestion
        if (JenkinsJVM.isJenkinsJVM()) {
```
They can do something: report a bug; uninstall the plugin; even adopt it. We will never catch long-tail usages based on absentee maintainers paying attention.

File: core/src/main/java/jenkins/SlaveToMasterFileCallable.java
Comment (jglick) [inline core/src/main/java/jenkins/SlaveToMasterFileCallable.java:41]: Or just throw a `SecurityException`, with a system property for an escape hatch that downgrades to the warning. Usages should be quite rare and we want to make them visible.

File: core/src/main/java/jenkins/SlaveToMasterFileCallable.java
Comment (daniel-beck) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: I think we can progressively restrict it further, no need to start with making plugins fail outright.

Comment (daniel-beck) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: > long-tail usages

S2MFCs can be statically identified, at least in Jenkins project and CloudBees hosted plugins. This isn't `FilePath` all over.

Comment (jglick) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: > making plugins fail outright

Was this for https://github.com/jenkinsci/jenkins/pull/5884#discussion_r743899465?

Comment (daniel-beck) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: Applies to both in a way? The possible progression is probably:

1. Deprecate only
2. Deprecate and be noisy during tests (this)
3. Deprecate and be noisy
4. Deprecate and fail stuff

Comment (jglick) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: Right, so I would recommend 3 or 4, but will defer to your judgment.

Comment (oleg-nenashev) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: I would prefer (3) but fine with (2)

**Review by oleg-nenashev — APPROVED** [review]
Fine with me. Note that, as a side effect, these messages will start appearing in Jenkinsfile Runner when `isUnitTest` is set. I am fine with that

**Comment by oleg-nenashev** [comment]
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

File: core/src/main/java/jenkins/SlaveToMasterFileCallable.java
Comment (basil) [reply core/src/main/java/jenkins/SlaveToMasterFileCallable.java:40]: > I would prefer (3) but fine with (2)

Ditto, for the reason given in [this comment](https://github.com/jenkinsci/jenkins/pull/5884/files#r743898577).

**Pull request merged**
